geode-core/src/main/java/org/apache/geode/pdx/internal/PeerTypeRegistration.java

```diff
@@ -365,20 +379,32 @@ public int defineType(PdxType newType) {
     }
     lock();
     try {
-      int id = getExistingIdForType(newType);
-      if (id != -1) {
-        return id;
+      if (typeToId.isEmpty() || typeToId.size() != idToType.size()) {
+        buildTypeToIdFromIdToType();
+      }
+      // flush the tmpTypeToId map
+      if (!tmpTypeToId.isEmpty()) {
+        typeToId.putAll(tmpTypeToId);
+        tmpTypeToId.clear();
+      }
+      // double check if my type is in region in case the typeToId map has been updated while
+      // waiting to obtain a lock
+      existingId = typeToId.get(newType);
+      if (existingId != null) {
+        return existingId;
       }
 
-      id = allocateTypeId(newType);
+      int id = allocateTypeId(newType);
       newType.setTypeId(id);
 
       updateIdToTypeRegion(newType);
-
-      typeToId.put(newType, id);
-
       return newType.getTypeId();
     } finally {
+      // flush the tmpTypeToId map for who introduced this new pdxType
+      if (!tmpTypeToId.isEmpty()) {
```

Comment:
Why is this finally needed?  How would the tmpTypeToId not already be empty here?

Reply:
We have discussed that, it's safe to do it when the creator holds the d-lock. 
The listener will only add into tmpTypeToId map now. We use it to make sure that the distribution should all finished before flushing anything into typeToId map (they stayed in tmpTypeToId map now)

Reply:
But wouldn't the line above (line 386) when we actually use the map, have cleared it?  Why do we need to reclear in the finally?  Why is the duplicated code needed in the finally block?  Same goes for the other method...